### ISSUE TITLE
Duplicate Charges

### ISSUE THREAD

**Issue by codyjames** [craftcms/commerce-stripe#207]
### Description
We're seeing duplicate charges occasionally. I'm unable to reproduce it. It doesn't have to do with pressing the submit button multiple times (we disable it after the first click). I'm wondering if this plugin is properly passing an idempotency_key.

https://stripe.com/docs/api/idempotent_requests

<img width="603" alt="Screen Shot 2022-05-18 at 1 37 33 PM" src="https://user-images.githubusercontent.com/378665/169151553-69c0813e-7427-4cd7-b742-83cf4e5d8163.png">

**Comment by jonleverrier** [comment]
Same scenario @codyjames. I've just posted this https://github.com/craftcms/commerce-stripe/issues/256. Have you seen any further duplicate charges?

**Comment by codyjames** [comment]
@jonleverrier Yeah, we get them occasionally and just have to refund the buyer. We had one with four charges recently.

**Comment by jonleverrier** [comment]
@codyjames Four charges!!! Were the timestamps for the 4 charges exactly the same? It's bizarre this kind of thing can happen. 

I'm thinking about writing a module that will 

- listen in on a payment event
- check the number of successful stripe payments
- if more than one successful card transaction, reject the other transactions?

**Comment by localcraig** [comment]
Is there any updates on this @lukeholder we've had the same issue. 
Customer has been charged 3 times.

**Comment by jaydensmith** [comment]
@lukeholder we've just had 2 instances of customers being charged twice.

- We are using a custom payment form (not `paymentFormHtml`).
- Stripe API version: 2020-08-27
- Not overriding mutex, but using Redis for sessions
- Hosted on Ubuntu 20.04.3 LTS, MariaDB 10.6.3
- Craft 4.6.0
- Commerce 4.4.0
- We are using a few commerce events, but nothing that mutates the order in a significant way.
`LineItems::EVENT_POPULATE_LINE_ITEM` - dynamically sets the line item price
`Order::EVENT_BEFORE_SAVE` - sets a single field on the order based on the line item contents
`Order::EVENT_AFTER_ORDER_PAID` - processes a Stripe Connect payout

![Screenshot 2024-01-22 at 10 47 30 am](https://github.com/craftcms/commerce-stripe/assets/16344913/0ab6f95d-5869-4254-9717-81e8f75b42d4)
![Screenshot 2024-01-22 at 10 47 44 am](https://github.com/craftcms/commerce-stripe/assets/16344913/182cf943-bb81-4e58-a186-79a704ce113e)

**Comment by jaydensmith** [comment]
This happened to us again today @lukeholder 
Pretty embarrassing to explain to customers.

**Comment by angrybrad** [comment]
@jaydensmith what hosting provider are you using?

**Comment by jaydensmith** [comment]
> @jaydensmith what hosting provider are you using?
> 
> Guessing based on this you're managing your own infrastructure?
> 
> > Hosted on Ubuntu 20.04.3 LTS, MariaDB 10.6.3

It's managed via Laravel Forge

**Comment by angrybrad** [comment]
@jaydensmith can you share your `config/app.php` file (redacting any sensitive bits of course)

**Comment by angrybrad** [comment]
@jaydensmith thanks for that.

Usually, when we see sporadic behaviors like this, the culprit has been using Redis as a mutex, which, as it turns out, doesn’t make a reliable distributed locking mechanism. https://redis.io/docs/manual/patterns/distributed-locks/

That doesn’t seem to be the case here, though.

I’m not sure what’s going on in your custom modules - it’s possible something in there is causing the behavior.

Guessing there’s no database read/write splitting based on the contents of `config/app.php`?

**Comment by angrybrad** [comment]
@jaydensmith I suppose as a next step, I'd update Craft and Commerce to the latest 4.x release to remove outdated versions as variables.

**Comment by codyjames** [comment]
I have a hunch that this is related to failed payment intents. It seems like this happens when a customer has some sort of issue with their payment. Then they resolve the issue and a new payment intent is created and successfully processed, but perhaps some of the old payment intents also get re-processed. Or perhaps the errors happened on the Craft side, and blocked the payment intent from being processed, and then once the order is able to go through without errors, Craft just processes all of the payment intents.

**Comment by lukeholder** [comment]
@codyjames yeah this has been hard for us to track down. Any steps you can determine to reproduce would be helpful.

**Comment by codyjames** [comment]
@lukeholder I wonder if the idempotency key needs to be re-thought. It feels like those should also help prevent multiple charges.

**Comment by lukeholder** [comment]
@codyjames thanks we are looking into improving this by only allowing a single payment intent per order, but will keep this ticket open until we find a solution. Thanks

**Comment by lukeholder** [comment]
Made a fix to web-hook payment intent success handling (6c4c1f114a14a22bb478509bfd306d4d6df30135) that I am hoping that this will stop people making a second duplicate payment since the order will be completed successfully on the first payment. Will update you here once the release is out.

**Comment by codyjames** [comment]
@lukeholder Is this a problem only for subscriptions? It's happening on non-subscriptions for us. I see your change was in the SubscriptionGateway file.
